src/task.c

```diff
@@ -737,6 +737,7 @@ JL_DLLEXPORT JL_NORETURN void jl_no_exc_handler(jl_value_t *e, jl_task_t *ct)
     if (eh != NULL) {                                                          \
         if (altstack) ptls->sig_exception = NULL;                              \
         pop_timings_stack()                                                    \
+        jl_eh_restore_state(eh);                                               \
```

Comment:
I assumed this had to run after the longjmp, as otherwise the state of the stack would be invalid afterwards (particularly for things like Windows StackOverflow). But maybe we already dealt with that issue elsewhere (in the implicit catch handling)?

Reply:
I don't think so? Nothing in this function touches the stack, only the current task state and as far as I know, we don't ever rely on it to do any sort of implicit sigreturn (which we do explicitly before entering this function), so there should be no difference. But of course, I may have overlooked something.

Reply:
This is not legal until after the longjmp as it touches the stack:
https://github.com/JuliaLang/julia/blob/85e3da02bbf0f8fe40166c1bb06653821f7c9f3c/src/rtutils.c#L276

Nor is any of this okay (since it tries to run GC and finalizers):
https://github.com/JuliaLang/julia/blob/85e3da02bbf0f8fe40166c1bb06653821f7c9f3c/src/rtutils.c#L296-L305

Reply:
I am okay with making this `leave` an implicit part of the catch block for the IR (like JL_CATCH did before), but we should still emit it into the catch block portion of the code, not run it before the throw

Reply:
Ok, I can make that change. It's slightly annoying because codegen needs to split the critical edge, but not a huge problem.